src/python/hash-funksjoner/hash-funksjoner.md

```diff
@@ -0,0 +1,116 @@
+
+Regelen er enkel nok: Start med 12 sifre, og del dem opp slik at du jobber med ett og ett. Legg sammen siffer nummer 2, 4, 6, osv, og multipliser summen med 3. Så legger du sammen sifrene på plass 1, 3, 5, osv, og legger til det du har fra før. Se på tallet du nå har fått, og finn ut hvor mye du må legge til for å komme opp til nærmeste tier. Det er sjekksifferet ditt.
+
+## Eksempel {.tip}
+
+Jeg har valgt en tilfeldig bok fra bokhylla mi. EAN-13-koden er 9780062731029. La oss sjekke at den stemmer.
+
+Først regner vi ut 3 · (7 + 0 + 6 + 7 + 1 + 2) = 69 og 9 + 8 + 0 + 2 + 3 + 0 = 22. 69 + 22 = 91. Den nærmeste tieren er 100, så da er sjekksifferet 9 -- akkurat som det skal være.
+
+## Sjekkliste {.check}
+
+- [ ] Åpne IDLE og lag en ny fil. Lim inn den følgende koden:
+  ```python
+  def verify_barcode(digits):
+
+      # Regn ut i henhold til formelen
+      sum = 3 * (digits[1] + digits[3] + digits[5] + digits[7]
+                 + digits[9] + digits[11])
+      print(sum)
+      sum = sum + (digits[0] + digits[2] + digits[4] + digits[6]
+                   + digits[8] + digits[10])
+
+      # Finn differansen mellom 10 og det siste sifferet i summen
+      # Sjekksifferet kan ikke være 10, så den siste % 10 gjør at det
+      # blir 0 i stedet.
+      checksum = (10 - (sum % 10)) % 10
+
+      if checksum == digits[12]:
+          return True
+      else:
+          return False
+  ```
+- [ ] Lagre og kjør koden med F5. I konsollen, test koden din med følgende kode: `verify_barcode([9,7,8,0,0,6,2,7,3,1,0,2,9])`.
+- [ ] Test koden din med andre EAN-koder. Se hva som skjer dersom du skriver et galt siffer, eller bytter om rekkefølgen på to.
+- [ ] Det er litt slitsomt å skrive inn EAN-koden siffer for siffer. Legg til den følgende koden i fila di:
+  ```python
+  def interpret_barcode(barcode):
+
+    # Vi sjekker først at vi faktisk har fått nøyaktig 13 sifre
+    if (len(barcode) != 13) or not (barcode.isdigit()):
+        return None
+    
+    # Oversett strengen til en liste av sifre
+    digits = [int(d) for d in barcode]
+
+    return digits
+  ```
+  Lagre filen. Du kan nå teste den ved å skrive `verify_barcode(interpret_barcode("9780062731029"))`
+  
+# Steg 2: Sterkere koder {.activity}
+
+Sjekksifferet i EAN-13 er godt nok så lenge det handler om å fange opp de feilene som gjøres ved et uhell, men det er ikke spesielt vanskelig å finne to strekkoder som har samme sjekksiffer. For en god hash-funksjon må det (blant annet) være vanskelig å finne to input som gir samme output.
```

Comment:
eksempel på to strekkoder med samme sjekksiffer? kan gi økt forståelse av å se at dersom to sifre i oddetall- eller partall-posisioner bytter plass så får man samme sjekksiffer osv

Reply:
Resolved.